README.md
# Web Crypto Tools

<p>
  <a
    href="https://github.com/willgm/web-crypto-tools/actions"
    target="_blank"
  >
    <img
      alt="Build"
      src="https://img.shields.io/github/workflow/status/willgm/web-crypto-tools/CI"
    />
  </a>
  <a
    href="https://www.npmjs.com/package/@webcrypto/tools"
    target="_blank"
  >
    <img
      alt="Version"
      src="https://img.shields.io/github/package-json/v/willgm/web-crypto-tools"
    />
  </a>
  <a
    href="https://github.com/willgm/web-crypto-tools/blob/master/LICENSE"
    target="_blank"
  >
    <img
      src="https://img.shields.io/badge/license-MIT-blue.svg"
      alt="web-crypto-tools is released under the MIT license"
    />
  </a>
  <a
    href="https://github.com/willgm/web-crypto-tools/graphs/contributors"
    target="_blank"
  >
    <img
      alt="Contributors"
      src="https://img.shields.io/github/contributors/willgm/web-crypto-tools.svg"
    />
  </a>
</p>

> This project is a set of tools to facilitate and give good defaults for use of the native **[Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API)**.

This project depends on the browser implementation of [Crypto API](https://caniuse.com/#feat=cryptography) and [TextEncoder API](https://caniuse.com/#feat=textencoder), which are both current implemented on all green browsers. If you do need to support IE or any older browser, you should look for available polyfills.

The native browser implementation of crypto algorithms are much more fast and secure than any other JavaScript library out there. But at the same time, it is a low level API that relies on you to decide every little detail of it, so this project will give you good defaults and a better developer experience, and still let you decide if you prefer use other algorithms or extra protections. Be aware that, even if this project facilitates the use of the Web Crypto API, it will not prevent you from make any mistakes if you have no idea about cryptography concepts, so take your time to study a little before use it in a real project.

In the end, this is a simple collection of stateless functions, values and types, that can be individually imported and used. The minified project has currently only about 3kb in total and it is also tree-shaking friendly, so you can end up using even less.

## :gear: Usage

### Install it at your project

```bash
npm install @webcrypto/tools --save
```

### Encrypt everything

```ts
import {
  generateBaseCryptoKey,
  deriveCryptKey,
  generateSalt,
  encryptValue,
  decryptValue,
  decode,
} from '@webcrypto/tools';

// get any data, string or typed arrays
const originalData = 'any data';

// create a secure base key that cannot be reverted to the original key value
const baseKey = await generateBaseCryptoKey('any raw key');

// create new keys for each crypto operation from the base key
const cryptoKey = await deriveCryptKey(baseKey, generateSalt());

// encrypt any value with military level security
const [cryptoValue, nonce] = await encryptValue(originalData, cryptoKey);

// decrypt your value when necessary
const decryptedValue = await decryptValue(cryptoValue, cryptoKey, nonce);

// the decrypted value should be the same of the original
expect(originalData).toEqual(decode(decryptedValue));
```

## :book: Documentation

The [documentation with all available API and options](https://willgm.github.io/web-crypto-tools/) at our GitHub Pages.

The [test cases](https://github.com/willgm/web-crypto-tools/tree/master/test) are also quite readable and can be used as example for all the possible API uses.

## License

[MIT](https://github.com/willgm/web-crypto-tools/blob/master/LICENSE)


CHANGELOG.md
# [2.0.0](https://github.com/willgm/web-crypto-tools/compare/v1.2.1...v2.0.0) (2022-02-17)


### Features

* support for TypeScript 4.5 ([998d68b](https://github.com/willgm/web-crypto-tools/commit/998d68b10b6c2eb02d428b3b011fcd72d0d7ea3a))


### BREAKING CHANGES

* Some type definitions on the public API needed to be changed due to TypeScript
breaking changes

## [1.2.1](https://github.com/willgm/web-crypto-tools/compare/v1.2.0...v1.2.1) (2022-02-16)


### Bug Fixes

* update all non breaking dependencies ([01c7c0e](https://github.com/willgm/web-crypto-tools/commit/01c7c0e5d927eb0aaef9d5d626a2c7f6b9c679cc))

# [1.2.0](https://github.com/willgm/web-crypto-tools/compare/v1.1.0...v1.2.0) (2020-10-26)


### Features

* **getCryptoObject:** method to get the crypto object depending on the environment ([bac4919](https://github.com/willgm/web-crypto-tools/commit/bac4919bd9f1f85c012f2267498972d00897172a))

# [1.1.0](https://github.com/willgm/web-crypto-tools/compare/v1.0.6...v1.1.0) (2020-05-29)


### Features

* **jwk:** adds support for importing key from JSON Web Key ([befcf93](https://github.com/willgm/web-crypto-tools/commit/befcf93dfc965c4dc801dd577689cf694b35ffa5))

## [1.0.6](https://github.com/willgm/web-crypto-tools/compare/v1.0.5...v1.0.6) (2020-05-27)


### Bug Fixes

* spelling ([5e840ab](https://github.com/willgm/web-crypto-tools/commit/5e840ab24cfc9d97a96fb33f98ad663f15801f38))

## [1.0.5](https://github.com/willgm/web-crypto-tools/compare/v1.0.4...v1.0.5) (2020-05-26)


### Bug Fixes

* publishing the entire dist folder to npm ([d3a3d40](https://github.com/willgm/web-crypto-tools/commit/d3a3d4014ee302ab51b40484e3f96e74c71deaea))

## [1.0.4](https://github.com/willgm/web-crypto-tools/compare/v1.0.3...v1.0.4) (2020-05-26)


### Bug Fixes

* correct build path for publishing ([5249610](https://github.com/willgm/web-crypto-tools/commit/52496107a8d59b4ec1dbb8922d7c5617405b1e64))

## [1.0.3](https://github.com/willgm/web-crypto-tools/compare/v1.0.2...v1.0.3) (2020-05-26)


### Bug Fixes

* publishing README to npm ([d20d36d](https://github.com/willgm/web-crypto-tools/commit/d20d36d10a99646b10a4fb85d10a13ea162fe24c))

## [1.0.2](https://github.com/willgm/web-crypto-tools/compare/v1.0.1...v1.0.2) (2020-05-26)


### Bug Fixes

* add repository url ([e1de904](https://github.com/willgm/web-crypto-tools/commit/e1de904bb3e2e69f4418d6036be2d361fe6d80c2))

## [1.0.1](https://github.com/willgm/web-crypto-tools/compare/v1.0.0...v1.0.1) (2020-05-26)


### Bug Fixes

* correcting release on npm ([279d9ba](https://github.com/willgm/web-crypto-tools/commit/279d9bab1fef38c4a4f801010895b2fee1102525))

# 1.0.0 (2020-05-26)


### Features

* open project ([105b8a1](https://github.com/willgm/web-crypto-tools/commit/105b8a1eb6b3db29f14c8fbaae5c73b859c64204))


test/web-crypto-tools.test.ts
import {
  decode,
  decryptValue,
  deriveCryptKey,
  encode,
  encryptValue,
  generateBaseCryptoKey,
  generateHash,
  generateNonce,
  generateRandomValues,
  generateSalt,
  getCryptoObject,
  isTypedArray,
} from '../src/web-crypto-tools';

describe('Web Crypto Tools', () => {
  describe('Getting Crypto Object', () => {
    const mscrypto = { fake: 666 } as any;
    let originalCrypto: Crypto;

    beforeEach(() => {
      window.msCrypto = mscrypto;
      originalCrypto = window.crypto;
    });
    afterEach(() => {
      delete window.msCrypto;
      try {
        global.crypto = originalCrypto;
        // tslint:disable-next-line:no-empty
      } catch (e) {}
    });

    it('should always get crypto object by official draft property', async () => {
      const subject = getCryptoObject();
      expect(subject).toBe(originalCrypto);
    });

    it('should get crypto object from msCrypto property if the official property is not accessible', async () => {
      delete global.crypto;
      const subject = getCryptoObject();
      expect(subject).toBe(mscrypto);
    });
  });

  describe('Base Key Creation', () => {
    it('should not be extractable extractable', async () => {
      const subject = await generateBaseCryptoKey('any raw key');
      expect(subject.extractable).toBeFalse();
    });

    it('should accept typed arrays as key data', async () => {
      const subject = await generateBaseCryptoKey(new Uint8Array(8));
      expect(subject.extractable).toBeFalse();
    });

    it('should use PBKDF2 algorithm by default for the base', async () => {
      const subject = await generateBaseCryptoKey('any raw key');
      expect(subject.algorithm).toEqual({ name: 'PBKDF2' });
    });

    it('should be used just for derive a new key by default', async () => {
      const subject = await generateBaseCryptoKey('any raw key');
      expect(subject.usages).toEqual(['deriveKey']);
    });

    it('should work with other algorithms and usages', async () => {
      const subject = await generateBaseCryptoKey(
        generateRandomValues(16),
        'AES-GCM',
        ['encrypt', 'decrypt'],
        'raw',
      );
      expect(subject.algorithm).toEqual({ name: 'AES-GCM', length: 128 } as any);
    });

    it('should accept JSON Web Key', async () => {
      const jwkEcKey = {
        crv: 'P-384',
        d: 'wouCtU7Nw4E8_7n5C1-xBjB4xqSb_liZhYMsy8MGgxUny6Q8NCoH9xSiviwLFfK_',
        ext: true,
        key_ops: ['sign'],
        kty: 'EC',
        x: 'SzrRXmyI8VWFJg1dPUNbFcc9jZvjZEfH7ulKI1UkXAltd7RGWrcfFxqyGPcwu6AQ',
        y: 'hHUag3OvDzEr0uUQND4PXHQTXP5IDGdYhJhL-WLKjnGjQAw0rNGy5V29-aV-yseW',
      };
      const subject = await generateBaseCryptoKey(
        jwkEcKey,
        {
          name: 'ECDSA',
          namedCurve: 'P-384',
        },
        ['sign'],
      );
      expect(subject.algorithm).toEqual({ name: 'ECDSA', namedCurve: 'P-384' } as any);
    });
  });

  describe('Key Derivation', () => {
    it('should be able do derive a new key from a base one', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt());
      expect(subject).toBeDefined();
    });

    it('should be able do derive a new key with custom interaction number', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt(), 100);
      expect(subject.algorithm).toBeDefined();
    });

    it('should not be extractable extractable', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt());
      expect(subject.extractable).toBeFalse();
    });

    it('should be used for encrypt and decrypt by default', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt());
      expect(subject.usages).toEqual(['encrypt', 'decrypt']);
    });

    it('should be able to config the key usage', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt(), undefined, ['encrypt']);
      expect(subject.usages).toEqual(['encrypt']);
    });

    it('should use AES-GCM algorithm with length of 256 by default', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt());
      expect(subject.algorithm).toEqual({ name: 'AES-GCM', length: 256 } as any);
    });

    it('should be able to use other derive algorithm', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, {
        name: 'PBKDF2',
        salt: generateSalt(),
        iterations: 100,
        hash: 'SHA-1',
      });
      expect(subject).toBeDefined();
    });

    it('should be able to use other target encryption algorithm', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const subject = await deriveCryptKey(baseKey, generateSalt(), {
        name: 'AES-CBC',
        length: 256,
      });
      expect(subject.algorithm).toEqual({ name: 'AES-CBC', length: 256 } as any);
    });
  });

  describe('Type Guard to Typed Array', () => {
    it('should return true when testing against any Int Array Types', () => {
      expect(isTypedArray(new Int8Array(8))).toBeTrue();
      expect(isTypedArray(new Int16Array(16))).toBeTrue();
      expect(isTypedArray(new Int32Array(32))).toBeTrue();
    });

    it('should return true when testing against any Uint Array Types', () => {
      expect(isTypedArray(new Uint8Array(8))).toBeTrue();
      expect(isTypedArray(new Uint16Array(16))).toBeTrue();
      expect(isTypedArray(new Uint32Array(32))).toBeTrue();
      expect(isTypedArray(new Uint8ClampedArray(8))).toBeTrue();
    });

    it('should return true when testing against any Float Array Types', () => {
      expect(isTypedArray(new Float32Array(32))).toBeTrue();
      expect(isTypedArray(new Float64Array(64))).toBeTrue();
    });

    it('should return true when testing against any Data Views', () => {
      expect(isTypedArray(new DataView(new ArrayBuffer(16)))).toBeTrue();
    });

    it('should return true when testing against any Array Buffer', () => {
      expect(isTypedArray(new ArrayBuffer(16))).toBeTrue();
    });

    it('should return false when testing against any other primitive type', () => {
      expect(isTypedArray(666)).toBeFalse();
      expect(isTypedArray('batata')).toBeFalse();
    });

    it('should return false when testing against any other object', () => {
      expect(isTypedArray({})).toBeFalse();
      expect(isTypedArray(new Object())).toBeFalse();
    });

    it('should return false when testing against any non typed array', () => {
      expect(isTypedArray([1, 2, 3])).toBeFalse();
      expect(isTypedArray([true, 'eita', {}])).toBeFalse();
    });
  });

  describe('Encrypting Values', () => {
    it('should return the encrypted the given value and a nonce when using strings', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue, nonce] = await encryptValue('any data', cryptoKey);
      expect(cryptoValue).toBeInstanceOf(ArrayBuffer);
      expect(nonce).toBeInstanceOf(Uint8Array);
    });

    it('should return the encrypted the given value and a nonce when using any typed array', async () => {
      const originalData = new Uint8Array(8);
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue, nonce] = await encryptValue(originalData, cryptoKey);
      expect(cryptoValue).not.toEqual(originalData);
      expect(cryptoValue).toBeInstanceOf(ArrayBuffer);
      expect(nonce).toBeInstanceOf(Uint8Array);
    });

    it('should return the encrypted the given value and a nonce when a custom AES algorithm', async () => {
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt(), {
        name: 'AES-CBC',
        length: 256,
      });
      const customAlgorithm = { name: 'AES-CBC', iv: generateNonce() };
      const [cryptoValue, nonce] = await encryptValue('any data', cryptoKey, customAlgorithm);
      expect(cryptoValue).toBeInstanceOf(ArrayBuffer);
      expect(nonce).toBe(customAlgorithm.iv);
    });

    it('should return the encrypted the given value and a null nonce when a custom algorithm without iv', async () => {
      const cryptoKey = await window.crypto.subtle.generateKey(
        {
          name: 'RSA-OAEP',
          modulusLength: 2048,
          publicExponent: new Uint8Array([1, 0, 1]),
          hash: 'SHA-256',
        },
        true,
        ['encrypt', 'decrypt'],
      );
      const customAlgorithm = { name: 'RSA-OAEP' };
      const [cryptoValue, nonce] = await encryptValue(
        'any data',
        cryptoKey.publicKey,
        customAlgorithm,
      );
      expect(cryptoValue).toBeInstanceOf(ArrayBuffer);
      expect(nonce).toBeNull();
    });
  });

  describe('Decrypting Values', () => {
    it('should decrypt the given value using the original nonce', async () => {
      const originalData = 'any data';
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue, nonce] = await encryptValue(originalData, cryptoKey);
      const decryptedValue = await decryptValue(cryptoValue, cryptoKey, nonce);
      expect(originalData).toEqual(decode(decryptedValue));
    });

    it('should decrypt the given value using the original nonce when using any typed array', async () => {
      const originalData = new Uint8Array([1, 9, 69, 666]);
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue, nonce] = await encryptValue(originalData, cryptoKey);
      const decryptedValue = await decryptValue(cryptoValue, cryptoKey, nonce);
      expect(originalData).toEqual(new Uint8Array(decryptedValue));
    });

    it('should decrypt the given value with a custom algorithm', async () => {
      const originalData = 'any data';
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt(), {
        name: 'AES-CBC',
        length: 256,
      });
      const customAlgorithm = { name: 'AES-CBC', iv: generateNonce() };
      const [cryptoValue] = await encryptValue('any data', cryptoKey, customAlgorithm);
      const decryptedValue = await decryptValue(cryptoValue, cryptoKey, customAlgorithm);
      expect(originalData).toEqual(decode(decryptedValue));
    });

    it('should not decrypt the given value using incorrect nonce', async () => {
      const originalData = 'any data';
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue] = await encryptValue(originalData, cryptoKey);
      const decryptOperation = decryptValue(cryptoValue, cryptoKey, generateNonce());
      await expectAsync(decryptOperation).toBeRejectedWithError();
    });

    it('should not decrypt the given value using incorrect crypto key', async () => {
      const originalData = 'any data';
      const baseKey = await generateBaseCryptoKey('any raw key');
      const cryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const [cryptoValue, nonce] = await encryptValue(originalData, cryptoKey);
      const wrongCryptoKey = await deriveCryptKey(baseKey, generateSalt());
      const error: Error = await decryptValue(cryptoValue, wrongCryptoKey, nonce).catch(e => e);
      expect(error.name).toBe('OperationError');
    });
  });

  describe('Random Values', () => {
    it('should generate a random value as a typed array with 8 random bytes by default', () => {
      const subject = generateRandomValues();
      expect(subject).toBeInstanceOf(Uint8Array);
      expect(subject.length).toBe(8);
    });

    it('should generate random value with the given byte size', () => {
      const byteSize = 69;
      const subject = generateRandomValues(byteSize);
      expect(subject.length).toBe(byteSize);
    });

    it('should generate a nonce as a typed array with 16 random bytes by default', () => {
      const subject = generateNonce();
      expect(subject).toBeInstanceOf(Uint8Array);
      expect(subject.length).toBe(16);
    });

    it('should generate nonce with the given byte size', () => {
      const byteSize = 69;
      const subject = generateNonce(byteSize);
      expect(subject.length).toBe(byteSize);
    });

    it('should generate a salt as a typed array with 8 random bytes by default', () => {
      const subject = generateSalt();
      expect(subject).toBeInstanceOf(Uint8Array);
      expect(subject.length).toBe(8);
    });

    it('should generate salt with the given byte size', () => {
      const byteSize = 69;
      const subject = generateSalt(byteSize);
      expect(subject.length).toBe(byteSize);
    });
  });

  describe('Encoding / Decoding', () => {
    it('should encode a string to typed array', () => {
      expect(encode('any string')).toBeInstanceOf(Uint8Array);
    });

    it('should check if it is already a typed array and just return it', () => {
      const originalValue = [1, 5, 10];
      const typedArray = new Uint8Array(originalValue);
      expect(encode(typedArray)).toEqual(new Uint8Array(originalValue));
    });

    it('should decode a typed array to string', () => {
      expect(decode(new Uint8Array([1, 5, 10]))).toBeInstanceOf(String);
    });

    it('should check if it is already a string and just return it', () => {
      const data = 'any string';
      expect(decode(data)).toBe(data);
    });

    it('should be able to decode an encoded value and get the original value', () => {
      const data = 'any string';
      expect(decode(encode(data))).toBe(data);
    });
  });

  describe('Hash Creation', () => {
    it('should generate a hash in a typed array of the input value', async () => {
      const subject = await generateHash('any string');
      expect(subject).toBeInstanceOf(ArrayBuffer);
    });

    it('should generate a hash in a typed array of another typed array', async () => {
      const originalValue = new Uint8Array([1, 5, 10]);
      const subject = await generateHash(originalValue);
      expect(subject).toBeInstanceOf(ArrayBuffer);
      expect(new Uint8Array(subject)).not.toEqual(originalValue);
    });

    it('should generate the same hash for the same input', async () => {
      const originalValue = 'any string';
      const hash1 = await generateHash(originalValue);
      const hash2 = await generateHash(originalValue);
      expect(new Uint8Array(hash1)).toEqual(new Uint8Array(hash2));
    });

    it('should generate the same hash for the same input in different algorithms', async () => {
      const originalValue = 'any string';
      const hash1 = await generateHash(originalValue, 'SHA-1');
      const hash2 = await generateHash(originalValue, 'SHA-256');
      expect(new Uint8Array(hash1)).not.toEqual(new Uint8Array(hash2));
    });

    it('should generate different hash for different inputs', async () => {
      const hash1 = await generateHash('first input');
      const hash2 = await generateHash('second input');
      expect(new Uint8Array(hash1)).not.toEqual(new Uint8Array(hash2));
    });

    it('should be able to uso other hash algorithms', async () => {
      const subject = await generateHash('any string', { name: 'SHA-1' });
      expect(subject).toBeInstanceOf(ArrayBuffer);
    });
  });
});


karma.conf.js
// Karma configuration

module.exports = function(config) {
  config.set({
    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine', 'karma-typescript'],

    // list of files / patterns to load in the browser
    files: ['test/*.test.ts', 'src/*.ts'],

    // list of files / patterns to exclude
    exclude: [],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      '**/*.ts': 'karma-typescript', // *.tsx for React Jsx
    },

    // Config: https://www.npmjs.com/package/karma-typescript
    karmaTypescriptConfig: {
      exclude: ['tools'],
      coverageOptions: {
        threshold: {
          global: {
            statements: 100,
            branches: 100,
            functions: 100,
            lines: 100,
          },
        },
      },
      reports: {
        html: 'coverage',
        'text-summary': '',
      },
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['mocha', 'karma-typescript'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['ChromeHeadless'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity,
  });
};


src/web-crypto-tools.ts
/**
 * Import Key Algorithms at Web Crypto API
 */
export type ImportAlgorithm =
  | AlgorithmIdentifier
  | RsaHashedImportParams
  | EcKeyImportParams
  | HmacImportParams
  | AesKeyAlgorithm;

/**
 * Derive Key Algorithms at at Web Crypto API
 */
export type DeriveAlgorithm = AlgorithmIdentifier | EcdhKeyDeriveParams | HkdfParams | Pbkdf2Params;

/**
 * Derive Algorithms Params for Web Crypto API
 */
export type DerivedAlgorithmFor =
  | AlgorithmIdentifier
  | AesDerivedKeyParams
  | HmacImportParams
  | HkdfParams
  | Pbkdf2Params;

/**
 * Params for Encrypt / Decrypt Algorithms at Web Crypto API
 */
export type CryptoAlgorithm =
  | AlgorithmIdentifier
  | RsaOaepParams
  | AesCtrParams
  | AesCbcParams
  | AesGcmParams;

/**
 * Possible uses of Crypto Keys
 */
export type CryptoKeyUsage =
  | 'encrypt'
  | 'decrypt'
  | 'deriveKey'
  | 'deriveBits'
  | 'wrapKey'
  | 'sign'
  | 'verify'
  | 'unwrapKey';

/**
 * Default number of iterations used with PBKDF2 algorithm
 */
export const PBKDF2_ITERATIONS_DEFAULT: number = 50000;

/**
 * @internal
 */
declare global {
  /**
   * IE11 use a different global property.
   * @internal
   */
  var msCrypto: Crypto;
}

/**
 * Returns the crypto object depending on browser support.
 * IE11 has support for the Crypto API, but it is in a different global scope.
 *
 * @returns The Crypto object.
 */
export function getCryptoObject(): Crypto {
  return window.crypto || window.msCrypto; // for IE 11
}

/**
 * Creates a base Crypto Key from the original raw key, by default this base key
 * should just be used to protect the original key to be discovery,
 * and should not be used directly to any encrypt / decrypt algorithm.
 * The generated base crypto key should be used just to derive new ones,
 * that then will be used to encrypt / decrypt algorithms.
 *
 * @param rawKey The original key to start the encrypt process.
 * @param algorithm The algorithm used to import the key.
 * @param keyUsages The uses for the generated Crypto Key.
 * @param format Input format for the raw key.
 * @returns A promise with the base Crypto Key.
 */
export function generateBaseCryptoKey(
  rawKey: string | BufferSource | JsonWebKey,
  algorithm: ImportAlgorithm = 'PBKDF2',
  keyUsages: KeyUsage[] = ['deriveKey'],
  format: KeyFormat = 'raw',
): Promise<CryptoKey> {
  const isJwkKey = !isTypedArray(rawKey) && typeof rawKey === 'object';
  return Promise.resolve(
    isJwkKey
      ? getCryptoObject().subtle.importKey(
          'jwk',
          rawKey,
          algorithm,
          false, // the original value will not be extractable
          keyUsages,
        )
      : getCryptoObject().subtle.importKey(
          format as Exclude<KeyFormat, 'jwk'>,
          typeof rawKey === 'string' ? encode(rawKey) : rawKey,
          algorithm,
          false, // the original value will not be extractable
          keyUsages,
        ),
  );
}

/**
 * Derives a base Crypto Key to new one that can be used in encrypt / decrypt algorithms
 * or any other possible uses in `CryptoKeyUsage`.
 *
 * @param cryptoBaseKey The base Crypto Key to be derive.
 * @param salt The salt value to be used with the default `PBKDF2` derive algorithm.
 * @param iterations The number of iterations to be used with the default `PBKDF2` derive algorithm. Default value: `PBKDF2_ITERATIONS_DEFAULT`.
 * @param keyUsages The new uses of the new derive Crypto Key. Default value: `['encrypt', 'decrypt']`.
 * @returns A promise with the derived Crypto Key for other uses.
 */
export function deriveCryptKey(
  cryptoBaseKey: CryptoKey,
  salt: BufferSource,
  iterations?: number,
  keyUsages?: CryptoKeyUsage[],
): Promise<CryptoKey>;

/**
 * Derives a base Crypto Key to new one that can be used in encrypt / decrypt algorithms
 * or any other possible uses in `CryptoKeyUsage`.
 *
 * @param cryptoBaseKey The base Crypto Key to be derive.
 * @param salt The salt value to be used with the default `PBKDF2` derive algorithm.
 * @param algorithmFor The algorithm where the derived Crypto Key will be used. Default value: `{ name: 'AES-GCM', length: 256 }`.
 * @param keyUsages The new uses of the new derive Crypto Key. Default value: `['encrypt', 'decrypt']`.
 * @returns A promise with the derived Crypto Key for other uses.
 */
export function deriveCryptKey(
  cryptoBaseKey: CryptoKey,
  salt: BufferSource,
  algorithmFor?: DerivedAlgorithmFor,
  keyUsages?: CryptoKeyUsage[],
): Promise<CryptoKey>;

/**
 * Derives a base Crypto Key to new one that can be used in encrypt / decrypt algorithms
 * or any other possible uses in `CryptoKeyUsage`.
 *
 * @param cryptoBaseKey The base Crypto Key to be derive.
 * @param deriveAlgorithm The algorithm to be used when deriving the Crypto Key.
 * @param algorithmFor The algorithm where the derived Crypto Key will be used. Default value: `{ name: 'AES-GCM', length: 256 }`.
 * @param keyUsages The new uses of the new derive Crypto Key. Default value: `['encrypt', 'decrypt']`.
 * @returns A promise with the derived Crypto Key for other uses.
 */
export function deriveCryptKey(
  cryptoBaseKey: CryptoKey,
  deriveAlgorithm: DeriveAlgorithm,
  algorithmFor?: DerivedAlgorithmFor,
  keyUsages?: CryptoKeyUsage[],
): Promise<CryptoKey>;

export function deriveCryptKey(
  cryptoBaseKey: CryptoKey,
  deriveAlgorithmOrSalt: DeriveAlgorithm | BufferSource,
  algorithmForOrIterations: DerivedAlgorithmFor | number = PBKDF2_ITERATIONS_DEFAULT,
  keyUsages: CryptoKeyUsage[] = ['encrypt', 'decrypt'],
): Promise<CryptoKey> {
  const deriveAlgorithm = isTypedArray(deriveAlgorithmOrSalt)
    ? ({
        name: 'PBKDF2',
        hash: 'SHA-256',
        salt: deriveAlgorithmOrSalt,
        iterations:
          typeof algorithmForOrIterations === 'number'
            ? algorithmForOrIterations
            : PBKDF2_ITERATIONS_DEFAULT,
      } as Pbkdf2Params)
    : deriveAlgorithmOrSalt;

  // The derived key will be used to encrypt with AES by default.
  const algorithmFor =
    typeof algorithmForOrIterations === 'number'
      ? ({ name: 'AES-GCM', length: 256 } as AesDerivedKeyParams)
      : algorithmForOrIterations;

  return Promise.resolve(
    getCryptoObject().subtle.deriveKey(
      deriveAlgorithm,
      cryptoBaseKey,
      algorithmFor,
      false, // the original key will not be extractable
      keyUsages,
    ),
  );
}

/**
 * Type Guard to Typed Array.
 *
 * @param data Any data to be checked.
 * @returns Verify if the given data is a Typed Array.
 */
export function isTypedArray(data: unknown): data is BufferSource {
  return ArrayBuffer.isView(data) || data instanceof ArrayBuffer;
}

/**
 * Encrypt a value with the given Crypto Key and Algorithm
 *
 * @param data Value to be encrypted.
 * @param cryptoKey The Crypto Key to be used in encryption.
 * @param algorithm The algorithm to be used in encryption. Default to `AES-GCM`.
 * @returns A promise with the encrypted value and the used nonce, if used with the encryption algorithm.
 */
export function encryptValue(
  data: string | BufferSource,
  cryptoKey: CryptoKey,
  algorithm: CryptoAlgorithm = { name: 'AES-GCM', iv: generateNonce() } as AesGcmParams,
): Promise<[ArrayBuffer, BufferSource | null]> {
  return Promise.resolve(
    getCryptoObject().subtle.encrypt(algorithm, cryptoKey, encode(data)),
  ).then(cryptoValue => [
    cryptoValue,
    typeof algorithm === 'object' && 'iv' in algorithm ? algorithm.iv : null,
  ]);
}

/**
 * Decrypt a value with the given Crypto Key and Algorithm
 *
 * @param data Value to be encrypted.
 * @param cryptoKey The Crypto Key used in encryption.
 * @param nonceOrAlgorithm The nonce used for AES encryption or the custom algorithm.
 * @returns A promise with the decrypt value
 */
export function decryptValue(
  data: BufferSource,
  cryptoKey: CryptoKey,
  nonceOrAlgorithm: BufferSource | CryptoAlgorithm,
): Promise<ArrayBuffer> {
  const algorithm = isTypedArray(nonceOrAlgorithm)
    ? ({ name: 'AES-GCM', iv: nonceOrAlgorithm } as AesGcmParams)
    : nonceOrAlgorithm;
  return Promise.resolve(getCryptoObject().subtle.decrypt(algorithm, cryptoKey, data));
}

/**
 * Generates random value to be used as nonce with encryption algorithms.
 *
 * @param byteSize The byte size of the generated random value.
 * @returns The random value.
 */
export function generateNonce(byteSize = 16): Uint8Array {
  // We should generate at least 16 bytes
  // to allow for 2^128 possible variations.
  return generateRandomValues(byteSize);
}

/**
 * Generates random value to be used as salt with encryption algorithms.
 *
 * @param byteSize The byte size of the generated random value.
 * @returns The random value.
 */
export function generateSalt(byteSize = 8): Uint8Array {
  // We should generate at least 8 bytes
  // to allow for 2^64 possible variations.
  return generateRandomValues(byteSize);
}

/**
 * Generates random value as a typed array of `Uint8Array`.
 *
 * @param byteSize The byte size of the generated random value.
 * @returns The random value.
 */
export function generateRandomValues(byteSize = 8): Uint8Array {
  return getCryptoObject().getRandomValues(new Uint8Array(byteSize));
}

/**
 * Encode a string value to a Typed Array as `Uint8Array`.
 * If the given value is already a Typed Array, then the value will be returned without any transformation.
 *
 * @param data Value to be encoded.
 * @returns The transformed given value as a Typed Array.
 */
export function encode(data: string | BufferSource): BufferSource {
  return isTypedArray(data) ? data : new TextEncoder().encode(data);
}

/**
 * Decode a ArrayBuffer value to a string.
 * If the given value is already a string, then the value will be returned without any transformation.
 *
 * @param data Value to be decoded.
 * @returns The transformed given value as a string.
 */
export function decode(data: string | BufferSource): string {
  return typeof data === 'string' ? data : new TextDecoder('utf-8').decode(data);
}

/**
 * Generates a hash value for the given value.
 *
 * @param data Seed value to generate a hash.
 * @param algorithm The algorithm to be used when generating the hash.
 * @returns A promise containing the hash value.
 */
export function generateHash(
  data: string | BufferSource,
  algorithm: string | Algorithm = 'SHA-256',
): Promise<ArrayBuffer> {
  return Promise.resolve(getCryptoObject().subtle.digest(algorithm, encode(data)));
}
